phoenix-core/src/main/java/org/apache/phoenix/hbase/index/builder/IndexBuildManager.java

```diff
@@ -88,27 +111,102 @@ public void getIndexUpdates(ListMultimap<HTableInterfaceReference, Pair<Mutation
       IndexMetaData indexMetaData) throws Throwable {
     // notify the delegate that we have started processing a batch
     this.delegate.batchStarted(miniBatchOp, indexMetaData);
-
+    CachedLocalTable cachedLocalTable =
+            preScanAllRequiredRows(
+                    mutations,
+                    (PhoenixIndexMetaData)indexMetaData,
+                    this.regionCoprocessorEnvironment);
     // Avoid the Object overhead of the executor when it's not actually parallelizing anything.
     for (Mutation m : mutations) {
-      Collection<Pair<Mutation, byte[]>> updates = delegate.getIndexUpdate(m, indexMetaData);
+      Collection<Pair<Mutation, byte[]>> updates = delegate.getIndexUpdate(m, indexMetaData, cachedLocalTable);
       for (Pair<Mutation, byte[]> update : updates) {
         indexUpdates.put(new HTableInterfaceReference(new ImmutableBytesPtr(update.getSecond())), new Pair<>(update.getFirst(), m.getRow()));
       }
     }
   }
 
+  @VisibleForTesting
+  public static CachedLocalTable preScanAllRequiredRows(
+          Collection<? extends Mutation> dataTableMutationsWithSameRowKeyAndTimestamp,
+          final PhoenixIndexMetaData indexMetaData,
+          RegionCoprocessorEnvironment regionCoprocessorEnvironment) throws IOException {
+      List<IndexMaintainer> indexTableMaintainers = indexMetaData.getIndexMaintainers();
+      Set<KeyRange> keys = new HashSet<KeyRange>(dataTableMutationsWithSameRowKeyAndTimestamp.size());
+      for (Mutation mutation : dataTableMutationsWithSameRowKeyAndTimestamp) {
+          keys.add(PVarbinary.INSTANCE.getKeyRange(mutation.getRow()));
+      }
+
+      Set<ColumnReference> getterColumnReferences = Sets.newHashSet();
+      for (IndexMaintainer indexTableMaintainer : indexTableMaintainers)
+      {
+          getterColumnReferences.addAll(
+                  indexTableMaintainer.getAllColumns());
+
+      }
+
+      getterColumnReferences.add(new ColumnReference(
+              indexTableMaintainers.get(0).getDataEmptyKeyValueCF(),
+              indexTableMaintainers.get(0).getEmptyKeyValueQualifier()));
+
+      Scan scan = IndexManagementUtil.newLocalStateScan(
+              Collections.singletonList(getterColumnReferences));
+      ScanRanges scanRanges = ScanRanges.createPointLookup(new ArrayList<KeyRange>(keys));
+      scanRanges.initializeScan(scan);
+      scan.setFilter(new SkipScanFilter(scanRanges.getSkipScanFilter(), true));
+
+      if(indexMetaData.getReplayWrite() != null) {
+          long timestamp = getMaxTimestamp(dataTableMutationsWithSameRowKeyAndTimestamp);
+          scan.setTimeRange(0, timestamp);
+      }
+
```

Comment:
We can eliminate unnecessary for raw and all versions scans for regular operations by just adding the following else part :
else {
          scan.setRaw(false);
          scan.setMaxVersions(1);
      }

Reply:
@kadirozde , thank you for the suggestion, I would modify following your opinion.

Reply:
@kadirozde ， seems that I added your 
else {
scan.setRaw(false);
scan.setMaxVersions(1);
}
The ConcurrentMutationsIT.testDeleteRowAndUpsertValueAtSameTS1 is failed.